flux-core: add v0.68.0 -> v0.73.0

vsoch opened the pull request:
The long awaited PR! So many versions. They are coming out of my ears. :ear: 

Ping @grondo.

Commit:
Automated deployment to update package flux-core 2025-04-04

grondo reviewed (approved):
Woo hoo 🙌 Thanks!

alecbcs reviewed (dismissed):
Looks good to me. Thanks @vsoch!

alecbcs commented:
@spackbot run pipeline

alecbcs commented:
@vsoch @grondo it looks like there's still an issue with py-cffi during the build that's failing CI.

https://gitlab.spack.io/spack/spack/-/jobs/16084427#L7113

vsoch commented:
I can't give you any help there @alecbcs - I debugged this weeks ago, and fixed it by removing those types. You can see the build is successful [here](https://github.com/flux-framework/spack/pull/308). When that came into a PR [here](https://github.com/spack/spack/pull/49656), your requested change was to add all three. Functionally that shouldn't have done anything, but it doesn't work anymore. I can't reproduce this locally, there are several layers of _different_ spack bugs that get hit before that, so I can't offer any more insights. When I install flux-sched in just about any environment I either pip install or apt/yum install python cffi and I never see this issue, so I suspect it's something with spack and Python paths.

My successful build is [here](https://github.com/flux-framework/spack/actions/runs/14012075636/job/39233138674) if you'd like a comparison.

grondo commented:
Is there a way to get spack ci to upload the `config.log` from the build as an artifact? That might give a hint what's missing. Other than that, I'm not sure why a package that requires cffi wouldn't be able to find it at build time in spack?

vsoch commented:
Oh, rofl, I already had added that to our builds (I just forgot).

![image](https://github.com/user-attachments/assets/c0520406-2f45-4c55-9397-114bc4082a8b)

Here is the artifact:

https://github.com/flux-framework/spack/actions/runs/14292239440/artifacts/2889989495

vsoch commented:
```
configure:17595: result: ${PYTHON_EXEC_PREFIX}/lib/python3.13/site-packages
configure:17619: checking for cffi.__version_info__ >= (1,1) in python module cffi
Traceback (most recent call last):
  File "<string>", line 4, in <module>
    from distutils.version import StrictVersion as Version
ModuleNotFoundError: No module named 'distutils'

During handling of the above exception, another exception occurred:
```

grondo commented:
Ah, I had forgotten that for Python 3.12 we'll now need the `packaging` module. See:

https://github.com/flux-framework/flux-core/pull/5459

for details. Sorry!

Note: The fix in that PR is the best I could come up with at the time. Happy to take other suggestions that would not require a new dependency!

vsoch committed:
Add py-packaging

vsoch commented:
Latest flux-core shouldn't be trying to build flux-sched 0.36.1 - that has very different requirements for gcc and clang. 

![image](https://github.com/user-attachments/assets/ddfbdd78-583a-4433-9c2b-236d11657170)

If it was the py-packaging addition that did it, I'll try removing the "when" to see what the concretizer chooses.

vsoch committed:
Do not pin py-packaging

vsoch commented:
@alecbcs let us know what you'd like to try. I'm not familiar with the spack CI hairball and why flux-sched is building at such an old version.  https://gitlab.spack.io/spack/spack/-/jobs/16100298

vsoch commented:
Ping @alecbcs what would you like to try next? The CI is outside of my spack expertise.

tgamblin commented:
> Latest flux-core shouldn't be trying to build flux-sched 0.36.1

`flux-sched` is brought in by `exaworks`, which is in the `e4s` stack. `exaworks` depends on `flux-sched`, which depends on `flux-core`.

The Linux pipelines build with gcc 11, but `flux-sched` has:

```python
conflicts("%gcc@:11", when="@0.37:", msg="gcc version must be 12 or higher")
```

So it's building older `flux-sched`, but failing on some header *that* `flux-sched` includes from `flux-core`.  So it looks like either older `flux-sched` should be constrained to use older `flux-core` (which I don't think accomplishes what you want -- if you want newer `flux-core` to be built in CI), or that header in core should be fixed so that older `flux-sched` with newer `flux-core` works with gcc 11.

vsoch commented:
I can comment that having built flux-sched and flux-core in a ton of environments, I either have needed to pin flux-sched to (at most) 0.37.0 and then flux-core to ~ 0.68.0, _or_ use newer gcc and build the latest for both. Flux core seems pretty flexible to build with different versions of flux-sched - the breaks always happen with gcc/clang versions and sched.

Is there any reason we can't update e4s / exaworks? There has been a ton of work and bugfixes for flux-sched (and core) since those versions. Ping @trws for thoughts.

trws commented:
I think what we're missing is a constraint in the flux-sched package for what versions of core that version of sched will build against.  The core side should build regardless of the version of sched, because it doesn't actually build against sched, but not necessarily the reverse.  I don't have a version compatibility table, but for example `depends_on("flux-core@:0.68", when="@:0.37")` would constrain core to 68 when building sched 37 because of old gcc.

trws commented:
*sigh* I should have realized I just reiterated what @tgamblin was saying. Regardless, that's what's going on.  Sched isn't tested to build against newer core versions currently.

vsoch committed:
flux-sched: build older flux-core

flux sched 0.38 was the first that required gcc
version 12 or higher, and flux-core continued to
build for some time, but eventually added
features that we are now seeing break with
sched 0.37 and the latest flux. This conflicts
should ensure that older flux-sched, which
is being built by having an older compiler,
only builds with flux-core up to 0.68.

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

vsoch commented:
I added a conflicts statement to flux-sched to reflect our discussion (thanks for that @tgamblin @trws) and (I think there is delay receiving the notification?) but [all the jobs passed](https://gitlab.spack.io/spack/spack/-/pipelines/1054454/) about 5 minutes ago.

@alecbcs could you give another review when you get the chance?

vsoch commented:
woot! Thanks all.
